llvm/include/llvm/IR/Instruction.h

```diff
@@ -303,8 +303,14 @@ class Instruction : public User,
   /// Get the metadata of given kind attached to this Instruction.
   /// If the metadata is not found then return null.
   MDNode *getMetadata(unsigned KindID) const {
-    if (!hasMetadata()) return nullptr;
-    return getMetadataImpl(KindID);
+    // Handle 'dbg' as a special case since it is not stored in the hash table.
+    if (KindID == LLVMContext::MD_dbg) {
+      return DbgLoc.getAsMDNode();
+    }
+    if (hasMetadataOtherThanDebugLoc()) {
+      return getMetadataImpl(KindID);
```

Comment:
Does this call check the debug kind and later hasMetadata again?

Reply:
From `Instruction.h`:
```
  /// Return true if this instruction has any metadata attached to it.
  bool hasMetadata() const { return DbgLoc || Value::hasMetadata(); }

  /// Return true if this instruction has metadata attached to it other than a
  /// debug location.
  bool hasMetadataOtherThanDebugLoc() const { return Value::hasMetadata(); }
```

Reply:
Note that `Value::hasMetadata()` is a protected member, accessing a bit stored on Value that caches whether anything is in the DenseMap stored in the LLVMContext.

Reply:
With my changes applied `getMetadataImpl()` will not perform any additional `hasMetadata()` checks anymore.

Reply:
right. I missed the change in Value::getMetadata().